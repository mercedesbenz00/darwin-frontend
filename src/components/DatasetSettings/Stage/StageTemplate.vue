<template>
  <div
    class="stage"
    :class="{
      'stage--headerless': !$slots.header,
      'stage--footerless': !$slots.footer
    }"
  >
    <div
      v-if="$slots.header"
      class="stage__header"
    >
      <slot name="header" />
    </div>
    <div class="stage__body">
      <slot name="body" />
    </div>
    <div
      v-if="$slots.footer"
      class="stage__footer"
    >
      <slot name="footer" />
    </div>
  </div>
</template>

<script lang="ts">
import { defineComponent } from 'vue'

export default defineComponent({ name: 'StageTemplate' })
</script>

<!-- eslint-disable vue-scoped-css/enforce-style-type -->
<style lang="scss">
:root {
  --stage-template-card-width: 215px;
  --stage-template-card-height: 393px;
  --stage-template-card-footer-height: 36px;
  --stage-template-card-header-height: auto;
}
</style>
<style lang="scss" scoped>

$boxBorder: 2px solid $colorAliceShadow;
$boxBorderRadius: 5px;

.stage {
  width: var(--stage-template-card-width);
  height: var(--stage-template-card-height);
  box-sizing: border-box;

  display: grid;
  grid-template-rows: var(--stage-template-card-header-height) 1fr var(--stage-template-card-footer-height);
  align-items: stretch;

  position: relative;

  border-radius: $boxBorderRadius;
  border: $boxBorder;
  background: $colorAliceBlue;
}

.stage__header,
.stage__footer {
  padding: 0 8px;
  align-items: center;
}

.stage__body {
  padding: 5px;
}

.stage__header {
  display: grid;
  align-items: center;
  column-gap: 5px;

  border-bottom: $boxBorder;
  grid-template-columns: auto 1fr auto auto;

  // We want headers for all stages to be of same height.
  // We also want that height to grow when we hover the editable title.
  // That means the height of the first row in the grid must be auto.
  // Adding padding to .stage__header itself would cause issues wit the rates
  // input, since that one is 22 px and adding 8 padding to that would mean
  // annotate and review stages could grow to up to 40px, while the other stages
  // remain at 37.
  // Adding margin to [contenteditable] instead ensures the header will only
  // grow if the [contenteditable grows]
  [contenteditable] {
    margin: 8px 0;
  }
}

.stage__body {
  // prevents grid blowout on horizontal overflow
  min-width: 0px;
}

.stage__footer {
  border-top: $boxBorder;
  display: grid;
  grid-template-columns: auto auto;
  justify-content: space-between;
}

.stage.stage--headerless {
  grid-template-rows: 1fr var(--stage-template-card-footer-height);
}
.stage.stage--footerless {
  grid-template-rows: var(--stage-template-card-header-height) 1fr;
}

.stage.stage.stage--footerless.stage.stage--headerless {
  grid-template-rows: 1fr;;
}

</style>
