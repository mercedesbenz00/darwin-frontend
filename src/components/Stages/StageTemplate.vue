<template>
  <div
    class="stage"
    :class="{ 'stage--active': selected }"
    :style="{ '--highlightColor': themeAttrs.highlightColor }"
    @click.stop="$emit('click')"
  >
    <div
      class="stage__head"
      :class="{
        'stage__head--in-only': themeAttrs.edges[0] && !themeAttrs.edges[1],
        'stage__head--out-only': themeAttrs.edges[1] && !themeAttrs.edges[0]
      }"
    >
      <StageEdge
        class="stage__head__edge"
        v-if="themeAttrs.edges[0]"
        :scale="scale"
        edge-type="in"
        :type="stage.type"
        :stage-id="stage.id"
        name="default"
      />
      <div class="stage__head__content">
        <div class="stage__head__content__icon">
          <slot name="icon" />
        </div>
        <h3 class="stage__head__content__title">
          {{ stageName }}
        </h3>
        <ItemCount :stage-id="stage.id" />
      </div>
      <StageEdge
        v-if="themeAttrs.edges[1]"
        class="stage__head__edge"
        :scale="scale"
        edge-type="out"
        :type="stage.type"
        :stage-id="stage.id"
        :edge="stage.edges[0]"
        :name="'default'"
      />
    </div>
    <div
      class="stage__content"
      v-if="!!$slots['content']"
    >
      <slot name="content" />
    </div>
  </div>
</template>

<script lang="ts">
import { computed, defineComponent, onMounted, onUnmounted } from 'vue'

import { useStore } from '@/composables/useStore'
import { useWorkflowSceneStore } from '@/composables/useWorkflowSceneStore'
import { useEditedWorkflow } from '@/pinia/useEditedWorkflow'
import { V2WorkflowStagePayload } from '@/store/types'
import { stageTheme } from '@/utils/workflowStageTheme'

import ItemCount from './ItemCount.vue'
import StageEdge from './StageEdge/StageEdge.vue'

export default defineComponent({
  name: 'StageTemplate',
  components: { ItemCount, StageEdge },
  props: {
    scale: { type: Number, required: true },
    stage: { type: Object as () => V2WorkflowStagePayload, required: true },
    selected: { default: false, required: false, type: Boolean }
  },
  setup (props, { emit }) {
    const workflowScene = useWorkflowSceneStore()
    const store = useStore()
    const Z_INDEX_SELECTED = 1005
    const Z_INDEX_DEFAULT = 1004

    const stageHeight = 'auto'
    const stageWidth = 240

    const zoomScale = computed(() => workflowScene.zoomScale)

    const { updateStageConfig } = useEditedWorkflow()

    const onDrag = (x: number, y: number): void => {
      const roundedX = Math.round(x)
      const roundedY = Math.round(y)

      updateStageConfig({
        stageId: props.stage.id,
        config: {
          x: roundedX,
          y: roundedY
        }
      })

      emit('on-drag', { x: roundedX, y: roundedY })
    }

    const unplugStage = (args: { stageId: string }): void => {
      store.commit('v2Workflow/UNPLUG_STAGE', args.stageId)
    }

    const deleteStage = (args: { stageId: string }): void => {
      unplugStage(args)
      store.commit('v2Workflow/REMOVE_STAGE', args.stageId)
      workflowScene.selectStage(null)
    }

    const onKeyDown = (event: KeyboardEvent): void => {
      if (event.target === null) { return }
      if (event.target.localName === 'input') { return }
      if ((event.key === 'Backspace' || event.key === 'Delete') && props.selected) {
        deleteStage({ stageId: props.stage.id })
      }
    }

    onMounted(() => {
      window.addEventListener('keydown', onKeyDown)
    })

    onUnmounted(() => {
      window.removeEventListener('keydown', onKeyDown)
    })

    const themeAttrs = computed(() => stageTheme[props.stage.type])
    const stageName = computed(() => props.stage.name || themeAttrs.value.stageTitle)

    return {
      onDrag,
      onKeyDown,
      stageName,
      stageHeight,
      stageWidth,
      themeAttrs,
      zoomScale,
      Z_INDEX_DEFAULT,
      Z_INDEX_SELECTED
    }
  }
})
</script>

<style lang="scss" scoped>
@import "@/components/Stages/assets/styles/mixins.scss";

.stage {
  display: block;
  max-height: 435px;
  background: $colorNeutralsLightWhite;
  box-shadow: $shadowLightXS;
  border-radius: 12px;
  border: none;
  z-index: var(--z-index-stage-container);

  &--active {
    &::after {
      content: '';
      display: block;
      position: absolute;
      top: -2px;
      left: -2px;
      width: 100%;
      height: 100%;
      border-radius: 12px;
      border: 2px solid var(--highlightColor);
      box-shadow: 0 2px 4px -1px rgba(49, 51, 53, 0.2);
      z-index: var(--z-index-stage-highlight);
      transition: all 225ms ease;
    }
  }

  &__head {
    position: relative;
    height: 100%;
    width: 100%;

    &__edge {
      position: absolute;
      top: calc(50% - 7px);

      &:first-child {
        left: -7px;
      }

      &:last-child {
        right: -7px;
      }
    }

    &__content {
      position: relative;
      flex: 1;
      display: grid;
      align-items: center;
      grid-template-columns: 20px 1fr auto;
      grid-template-rows: 1fr;
      grid-gap: 4px;
      width: 100%;
      padding: 10px;

      &__icon {
        display: flex;
        align-items: center;
        justify-content: center;
        width: 20px;
        height: 20px;
      }

      &__title {
        user-select: none;
        max-width: 100%;
        @include ellipsis(1, md-1);
        @include typography(md-1, inter, 500);
      }
    }
  }

  &__content {
    display: block;
    height: calc(100% - 81px);
    border-top: 1px solid $colorNeutralsLight300;
  }
}

</style>
