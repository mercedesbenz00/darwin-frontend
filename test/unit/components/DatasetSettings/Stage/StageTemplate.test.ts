import { createLocalVue, shallowMount } from '@vue/test-utils'

import StageTemplate from '@/components/DatasetSettings/Stage/StageTemplate.vue'

const localVue = createLocalVue()

let slots: {
  body: string
} | {
  header: string
  body: string
} | {
  body: string
  footer: string
} | {
  header: string
  body: string
  footer: string
}

const itMatchesSnapshot = () => {
  it('matches snapshot', () => {
    const wrapper = shallowMount(StageTemplate, { localVue, slots })
    expect(wrapper).toMatchSnapshot()
  })
}

describe('with header', () => {
  beforeEach(() => {
    slots = {
      header: 'Header',
      body: 'Body'
    }
  })

  itMatchesSnapshot()

  it('renders slots', () => {
    const wrapper = shallowMount(StageTemplate, { localVue, slots })
    expect(wrapper.text()).toContain('Header')
    expect(wrapper.text()).toContain('Body')
  })
})

describe('with footer', () => {
  beforeEach(() => {
    slots = {
      body: 'Body',
      footer: 'Footer'
    }
  })

  itMatchesSnapshot()

  it('renders slots', () => {
    const wrapper = shallowMount(StageTemplate, { localVue, slots })
    expect(wrapper.text()).toContain('Body')
    expect(wrapper.text()).toContain('Footer')
  })
})

describe('with header and footer', () => {
  beforeEach(() => {
    slots = {
      header: 'Header',
      body: 'Body',
      footer: 'Footer'
    }
  })

  itMatchesSnapshot()

  it('renders slots', () => {
    const wrapper = shallowMount(StageTemplate, { localVue, slots })
    expect(wrapper.text()).toContain('Header')
    expect(wrapper.text()).toContain('Body')
    expect(wrapper.text()).toContain('Footer')
  })
})

describe('without header and footer', () => {
  beforeEach(() => {
    slots = {
      body: 'Body'
    }
  })

  itMatchesSnapshot()

  it('renders slots', () => {
    const wrapper = shallowMount(StageTemplate, { localVue, slots })
    expect(wrapper.text()).toContain('Body')
  })
})
