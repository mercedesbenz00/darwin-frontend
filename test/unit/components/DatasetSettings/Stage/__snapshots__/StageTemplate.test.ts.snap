// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`with footer matches snapshot 1`] = `
<div class="stage stage--headerless">
  <!---->
  <div class="stage__body">Body</div>
  <div class="stage__footer">Footer</div>
</div>
`;

exports[`with header and footer matches snapshot 1`] = `
<div class="stage">
  <div class="stage__header">Header</div>
  <div class="stage__body">Body</div>
  <div class="stage__footer">Footer</div>
</div>
`;

exports[`with header matches snapshot 1`] = `
<div class="stage stage--footerless">
  <div class="stage__header">Header</div>
  <div class="stage__body">Body</div>
  <!---->
</div>
`;

exports[`without header and footer matches snapshot 1`] = `
<div class="stage stage--headerless stage--footerless">
  <!---->
  <div class="stage__body">Body</div>
  <!---->
</div>
`;
